daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Equality.scala

```diff
@@ -16,60 +12,81 @@ private[lf] object Equality {
-          case (SRecord(tyCon1, fields1, args1), SRecord(tyCon2, fields2, args2)) =>
-            tyCon1 == tyCon2 && (fields1 sameElements fields2) &&
-              equality(zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack))
-          case (SVariant(tyCon1, _, rank1, arg1), SVariant(tyCon2, _, rank2, arg2)) =>
-            tyCon1 == tyCon2 && rank1 == rank2 && equality((arg1, arg2) +: stack)
-          case (SList(lst1), SList(lst2)) =>
-            lst1.length == lst2.length &&
-              equality(zipAndPush(lst1.iterator, lst2.iterator, stack))
-          case (SOptional(None), SOptional(None)) =>
-            equality(stack)
-          case (SOptional(Some(v1)), SOptional(Some(v2))) =>
-            equality((v1, v2) +: stack)
-          case (STextMap(map1), STextMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SGenMap(map1), SGenMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SStruct(fields1, args1), SStruct(fields2, args2)) =>
-            (fields1 sameElements fields2) && equality(
-              zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack),
-            )
-          case (SAny(t1, v1), SAny(t2, v2)) =>
-            t1 == t2 && equality((v1, v2) +: stack)
-          case (STypeRep(t1), STypeRep(t2)) =>
-            t1 == t2 && equality(stack)
-          case _ =>
-            false
-        }
-      case _ =>
+    while (success && stack.nonEmpty) success = stack.pop() match {
+      case (x: SPrimLit, y: SPrimLit) =>
+        x == y
+      case (SEnum(_, _, rank1), SEnum(_, _, rank2)) =>
+        rank1 == rank2
+      case (SRecord(_, _, args1), SRecord(_, _, args2)) =>
+        stack.push(args1.iterator().asScala, args2.iterator().asScala)
         true
+      case (SVariant(_, _, rank1, arg1), SVariant(_, _, rank2, arg2)) =>
+        stack.push(arg1, arg2)
+        rank1 == rank2
+      case (SList(lst1), SList(lst2)) =>
+        stack.push(lst1.iterator, lst2.iterator)
+        lst1.length == lst2.length
+      case (SOptional(xs), SOptional(ys)) =>
+        stack.push(xs.iterator, ys.iterator)
+        xs.nonEmpty == ys.nonEmpty
+      case (STextMap(map1), STextMap(map2)) =>
+        val keys = map1.keys
+        stack.push(keys.iterator.map(map1), keys.iterator.map(map2))
+        map1.size == map2.size && keys.forall(map2.keySet.contains)
+      case (SGenMap(map1), SGenMap(map2)) =>
+        stack.push(map1.values.iterator, map2.values.iterator)
+        stack.push(map1.keys.iterator, map2.keys.iterator)
+        map1.size == map2.size
+      case (SStruct(fields1, args1), SStruct(fields2, args2)) =>
```

Comment:
Have we decided if we want the fields to be sorted or not? I don't see any reason why we should identify the types `<a: S, b: T>` and `<b: T, a:S>` since we only use structural records for type class dictionaries, which cannot be compared.